api/models/datastore.go

```diff
@@ -59,6 +59,24 @@ type Datastore interface {
 	// ErrDatastoreEmptyRoutePath when routePath is empty. Returns ErrRoutesNotFound when no route exists.
 	RemoveRoute(ctx context.Context, appID, routePath string) error
 
+	// PutFunc inserts a new function if one does not exist, applying any defaults necessary, or
+	// updates a function that exists under the same name. Returns ErrDatastoreEmptyFunc if func is nil,
+	// ErrDatastoreEmptyFuncName is func.Name is empty.
+	// TODO(reed): should we allow rename if id provided?
+	PutFunc(ctx context.Context, fn *Func) (*Func, error)
+
+	// GetFuncs returns a list of funcs, applying any additional filters provided.
+	GetFuncs(ctx context.Context, filter *FuncFilter) ([]*Func, error)
+
+	// GetFunc returns a function by name. Returns ErrDatastoreEmptyFuncName if funcName is empty.
+	// Returns ErrFuncsNotFound if a func is not found.
+	// TODO(reed): figure out addressable by id or name biz. iff 1 query, name works.
+	GetFunc(ctx context.Context, funcName string) (*Func, error)
+
+	// RemoveFunc removes a function. Returns ErrDatastoreEmptyFuncName if funcName is empty.
+	// Returns ErrFuncsNotFound if a func is not found.
```

Comment:
FuncNotFound isn't an error... it's expected behaviour, maybe this is a goism i'm not used to.

Reply:
on delete (well, most endpoints) for all of our resources we return not found if it didn't exist. as far as go errors, we encode http statuses in these fancy errors https://github.com/fnproject/fn/blob/master/api/models/error.go as well as implementing `models.APIError` to hide errors we're less sure of from users.